framework/include/materials/MaterialPropertyInterface.h

```diff
@@ -681,3 +703,25 @@ MaterialPropertyInterface::hasADMaterialPropertyByName(const std::string & name_
                         : MooseUtils::join(std::vector<std::string>({name_in, _get_suffix}), "_");
   return _material_data->haveADProperty<T>(name);
 }
+
+template <typename T>
+const MaterialProperty<T> * const &
+MaterialPropertyInterface::getOptionalMaterialProperty(const std::string & name)
+{
+  _optional_material_pointer_storage.push_back(nullptr);
+  auto & new_pointer = _optional_material_pointer_storage.back();
+  if (hasMaterialProperty<T>(name))
+    new_pointer = &getMaterialProperty<T>(name);
+  return reinterpret_cast<const MaterialProperty<T> *&>(new_pointer);
+}
+
+template <typename T>
+const ADMaterialProperty<T> * const &
+MaterialPropertyInterface::getOptionalADMaterialProperty(const std::string & name)
+{
+  _optional_material_pointer_storage.push_back(nullptr);
+  auto & new_pointer = _optional_material_pointer_storage.back();
+  if (hasADMaterialProperty<T>(name))
+    new_pointer = &getADMaterialProperty<T>(name);
+  return reinterpret_cast<const ADMaterialProperty<T> *&>(new_pointer);
+}
```

Comment:
```suggestion
template <typename T, bool is_ad>
const GenericMaterialProperty<T> * const &
MaterialPropertyInterface::getOptionalMaterialProperty(const std::string & name)
{
    if (hasGenericMaterialProperty<T, is_ad>(name))
      return &getGenericMaterialProperty<T, is_ad>(name);
    static const GenericMaterialProperty<T, is_ad> * null_value = nullptr;
    return null_value;
}
```

ooo what about this?

Reply:
Ah damn.. disregard. Of course it won't; we still need to store a reference to a pointer. Oh well. I think this can be improved though with more "generic"

Reply:
Okay yeah, I say definitely go the generic route here. Up to you if you wanna use the static or not:
```
template <typename T, bool is_ad>
const GenericMaterialProperty<T, is_ad> * const &
MaterialPropertyInterface::getGenericOptionalMaterialProperty(const std::string & name)
{
  if (hasGenericMaterialProperty<T, is_ad>(name))
  {
    _optional_material_pointer_storage.push_back(&getGenericMaterialProperty<T, is_ad>(name));
    return reinterpret_cast<const GenericMaterialProperty<T, is_ad> *&>(
        _optional_material_pointer_storage.back());
  }
  static const GenericMaterialProperty<T, is_ad> * const null_value = nullptr;
  return null_value;
}
```

may be a bit excessive